plugins/logs/plugin.go

```diff
@@ -584,17 +584,22 @@ func (p *Plugin) oneShot(ctx context.Context) (ok bool, err error) {
 		return false, nil
 	}
 
+	var chunkUploadErr error
 	for bs := oldBuffer.Pop(); bs != nil; bs = oldBuffer.Pop() {
 		err := uploadChunk(ctx, p.manager.Client(p.config.Service), p.config.PartitionName, bs)
 		if err != nil {
 			// requeue the chunk
 			p.mtx.Lock()
 			p.bufferChunk(p.buffer, bs)
 			p.mtx.Unlock()
-			return false, err
+			chunkUploadErr = err
 		}
 	}
```

Comment:
❓ I might be missing something: Couldn't this, with multiple iterations of the for loop, be overwriting `chunkUploadError`? So, effectively, the error of the last failing `uploadChunk` invocation would be the one returned, not the first, or any in between.

This might not be an issue for reasons I don't seem glancing at this code snippet, though... 😅

Reply:
No, you are absolutely right in that :) My thinking was that the caller would be interested mainly in "did something go wrong?" and be able to log or act on that - not the specifics of how many times something failed. If you think it's better we could do something like incrementing a counter per error and then return an error with an aggregated message like "Failed uploading 5 chunks out of 10 in buffer, retrying at next interval" or something like that instead?

Reply:
Hmm I don't know. 😅 Any failure would cause a retry, correct? 🤔 So, the chance that multiple attempts fail for different reasons might be remote... hmm. Don't let this block you, I haven't thought much about the problem...

Reply:
Ok, thinking about this: if one of the chunks has failed to upload, does it even make sense to keep uploading the rest? Maybe we should stop attempting to when the first error is encountered...? Ah, any chunk that failed is pushed into the _new buffer_, to be re-pushed on the next attempt. 💡 In the process, stuff could fall on the floor, if the buffer is full; we'll end up discarding events. (Sorry this is verbose, I'm just taking notes as I work on understandin the plugin code 😅)

So, for oldBuffer, we'll go through all chunks, and end up either pushing them or forwarding them to the new buffer. If we fail, we take note of the error, and return the last upload error encountered. It'll end up being logged.

I guess we could use something like [`multierr`](https://pkg.go.dev/go.uber.org/multierr), but if we eshew the dependency for this edge case, I think I rather like your "Failed uploading 5 chunks out of 10 in buffer" idea. That said, what's in the PR now is providing more insight into what's happening already, so we might as well merge it as-is 👍

Reply:
I appreciate the notes @srenatus .. I think you're on to what I mentioned in the top comment:

> As for the fix itself - an alternative (and perhaps more sensible?) approach would be to stop uploading at first error but to push all remaining chunks in the old buffer into the new one to be preserved for next invocation.

It's probably unlikely that one chunk upload fails and the rest of them won't, but it could happen. In the common scenario we'll just wait for the next iteration (by default 300 seconds later, IIRC) to try and resend.. but I actually encountered this problem when trying to take a stab at issue #780, and in that scenario it might make more sense to try and "force" flushing the logs even if one chunk upload should fail. So I think both approaches are fine - and they're both definitely an improvement compared to having the logs discarded :D